addons/membership/models/partner.py

```diff
@@ -47,56 +47,38 @@ class Partner(models.Model):
-                if partner.free_member:
-                    partner.membership_state = 'free'
-                    continue
-            if partner.associate_member:
-                partner.associate_member._compute_membership_state()
-                partner.membership_state = partner.associate_member.membership_state
-                continue
+            if not partner.membership_state or (not partner.free_member and partner.membership_state == 'free'):
+                partner.membership_state = 'none'
 
             line_states = [mline.state for mline in partner.member_lines if \
                            (mline.date_to or date.min) >= today and \
                            (mline.date_from or date.min) <= today and \
                            mline.account_invoice_line.move_id.partner_id == partner]
 
             if 'paid' in line_states:
-                state = 'paid'
+                partner.membership_state = 'paid'
             elif 'invoiced' in line_states:
-                state = 'invoiced'
+                partner.membership_state = 'invoiced'
             elif 'waiting' in line_states:
-                state = 'waiting'
+                partner.membership_state = 'waiting'
             elif 'canceled' in line_states:
-                state = 'canceled'
-
-            if state == 'none':
+                partner.membership_state = 'canceled'
+            else:
                 for mline in partner.member_lines:
-                    # if there is an old invoice paid, set the state to 'old'
-                    if ((mline.date_from or date.min) < today and (mline.date_to or date.min) < today and \
-                            (mline.date_from or date.min) <= (mline.date_to or date.min) and \
-                            mline.account_invoice_id and mline.account_invoice_id.payment_state in ('in_payment', 'paid')):
-                        state = 'old'
+                    if ((mline.date_from or date.min) < today and (mline.date_to or date.min) <= today and \
+                            (mline.date_from or date.min) < (mline.date_to or date.min)):
+                        if mline.account_invoice_id and mline.account_invoice_id.payment_state in ('in_payment', 'paid'):
+                            partner.membership_state = 'old'
+                        elif mline.account_invoice_id and mline.account_invoice_id.state == 'cancel':
+                            partner.membership_state = 'canceled'
                         break
 
-            if partner.free_member and state != 'paid':
-                state = 'free'
-            partner.membership_state = state
+            if partner.free_member and partner.membership_state != 'paid':
+                partner.membership_state = 'free'
+
+            if partner.associate_member:
+                partner.associate_member._compute_membership_state()
+                partner.membership_state = partner.associate_member.membership_state
```

Comment:
in case of an `associate_member` the value of `membership_state` will be copied so it seems to be ok currently: we check early to possibly skip the rest
To improve this, adding the dependency `'associate_member.membership_state'` we can avoid the nested compute call